### ISSUE TITLE
result of LigerCrossEntropyLoss is always 0

### ISSUE THREAD

**Issue by wa008** [linkedin/Liger-Kernel#507]
### 🐛 Describe the bug

As title, code as below
```
from liger_kernel.transformers.cross_entropy import LigerCrossEntropyLoss

torch.manual_seed(2024)
M = 4
N = 3
input = torch.randn(M, N, requires_grad = False, device = DEVICE)
target = torch.empty(M, dtype = torch.long, device = DEVICE).random_(N)

liger_loss = LigerCrossEntropyLoss()
output_liger = liger_loss(input, target)

torch_loss_lib = torch.nn.CrossEntropyLoss()
output_torch_lib = torch_loss_lib(input, target)

print(f'input: {input}')
print(f'target: {target}')
print(f'output_liger: {output_liger}')
print(f'output_torch_lib: {output_torch_lib}')
```
output: 
```
input: tensor([[-0.1118,  0.0162,  0.0956],
        [-0.1134,  0.0088,  0.1046],
        [-0.1737,  0.1315,  0.0422],
        [-0.0921,  0.0728,  0.0194]], device='cuda:0')
target: tensor([0, 0, 0, 0], device='cuda:0')
output_liger: 0.0
output_torch_lib: 1.2258632183074951
```

### Reproduce

I run it on colab and install it by pip
```
!pip install liger-kernel
```

### Versions

Environment Report:
-------------------
Operating System: Linux-6.1.85+-x86_64-with-glibc2.35
Python version: 3.10.12
Liger Kernel version: 0.5.2
PyTorch version: 2.5.1+cu121
CUDA version: 12.1
HIP(ROCm) version: Not available
Triton version: 3.1.0
Transformers version: 4.47.1
XPU version: XPU Not Available

**Comment by DandinPower** [comment]
Hi, I think that because the Liger kernel is designed for memory optimization, the input tensor's storage space is used in-place to store the gradients during the forward computation. As a result, if you print the input before and after the `liger_loss` in your code, you'll see that the input tensor becomes the gradient values for the input. (You can verify this by setting `requires_grad=True` and printing `input.grad` after calling `liger_loss`.)

Given your test code, the output loss cannot be the same because the inputs to Liger CE and Torch CE are different.

The correct way to test this might look like the following:
```python
from liger_kernel.transformers.cross_entropy import LigerCrossEntropyLoss

torch.manual_seed(2024)
M = 4
N = 3
input = torch.randn(M, N, requires_grad=False, device=DEVICE)
target = torch.empty(M, dtype=torch.long, device=DEVICE).random_(N)

liger_input = input.detach().clone()
torch_input = input.detach().clone()

liger_loss = LigerCrossEntropyLoss()
output_liger = liger_loss(liger_input, target)

torch_loss_lib = torch.nn.CrossEntropyLoss()
output_torch_lib = torch_loss_lib(torch_input, target)

print(f'Input: {input}')
print(f'Target: {target}')
print(f'Output (Liger): {output_liger}')
print(f'Output (Torch): {output_torch_lib}')
```

However, this approach still doesn't explain why the output becomes `0.0`. I tested your code on my local 4090 machine and in a Colab T4 runtime, but I couldn't reproduce the result where the output is always `0.0`. Could you provide more details about your environment or any specific conditions so I can better try to reproduce your situation?

**Comment by Tcc0403** [comment]
Ditto. I slightly modified your code to show that two functions got different input, and I couldn't reproduce the result on colab either.
```python
from liger_kernel.transformers.cross_entropy import LigerCrossEntropyLoss
import torch
DEVICE = 'cuda'
torch.manual_seed(2024)
M = 4
N = 3
input = torch.randn(M, N, requires_grad = False, device = DEVICE)
target = torch.empty(M, dtype = torch.long, device = DEVICE).random_(N)
print(f'input: {input}')
print(f'target: {target}')
liger_loss = LigerCrossEntropyLoss()
output_liger = liger_loss(input, target)

print(f'output_liger: {output_liger}')

torch_loss_lib = torch.nn.CrossEntropyLoss()
print(f'input: {input}')
print(f'target: {target}')
output_torch_lib = torch_loss_lib(input, target)


print(f'output_torch_lib: {output_torch_lib}')
```
output 
```
input: tensor([[ 0.9979, -1.1449,  0.6296],
        [ 0.7755, -1.9720,  0.5091],
        [ 0.7670,  1.3114,  0.1737],
        [ 2.6832,  1.9087,  0.5860]], device='cuda:0')
target: tensor([0, 0, 0, 0], device='cuda:0')
output_liger: 0.7109375596046448
input: tensor([[-0.1118,  0.0162,  0.0956],
        [-0.1134,  0.0088,  0.1046],
        [-0.1737,  0.1315,  0.0422],
        [-0.0921,  0.0728,  0.0194]], device='cuda:0')
target: tensor([0, 0, 0, 0], device='cuda:0')
output_torch_lib: 1.2258632183074951
```

**Comment by wa008** [comment]
Really thanks for your nice reply. @Tcc0403 @DandinPower 
I found that is because I installed and set some environment variables for Triton, it will go wrong if I add below code.
```
import os
os.environ["TRITON_INTERPRET"] = "1"
```

full code: https://colab.research.google.com/drive/19-X6HpS9Xl5pnhJHg0LX4vB6FeTQI1Lf?usp=drive_link

**Comment by DandinPower** [comment]
This issue might arise because, when using Triton in interpreter mode, indirect memory access is not supported ([reference](https://triton-lang.org/main/programming-guide/chapter-3/debugging.html#limitations)). For example, patterns like the following are not supported:

```python
ptr = tl.load(ptr)
x = tl.load(ptr)
```

The current cross entropy kernel contains a few instances of this pattern. For example, in `ops/cross_entropy.py`, line 85 ([L85](https://github.com/linkedin/Liger-Kernel/blob/main/src/liger_kernel/ops/cross_entropy.py#L85)) has the statement `y = tl.load(Y_ptr)`, and line 109 ([L109](https://github.com/linkedin/Liger-Kernel/blob/main/src/liger_kernel/ops/cross_entropy.py#L109)) uses `ori_X_y = tl.load(X_ptr + y).cast(tl.float32)`.

**Comment by wa008** [comment]
> This issue might arise because, when using Triton in interpreter mode, indirect memory access is not supported ([reference](https://triton-lang.org/main/programming-guide/chapter-3/debugging.html#limitations)). For example, patterns like the following are not supported:
> 
> ```python
> ptr = tl.load(ptr)
> x = tl.load(ptr)
> ```
> 
> The current cross entropy kernel contains a few instances of this pattern. For example, in `ops/cross_entropy.py`, line 85 ([L85](https://github.com/linkedin/Liger-Kernel/blob/main/src/liger_kernel/ops/cross_entropy.py#L85)) has the statement `y = tl.load(Y_ptr)`, and line 109 ([L109](https://github.com/linkedin/Liger-Kernel/blob/main/src/liger_kernel/ops/cross_entropy.py#L109)) uses `ori_X_y = tl.load(X_ptr + y).cast(tl.float32)`.

Thanks again, I'm unsure I understand you correctly, I try below code which indirectly access memory and it works well.
```
    x = tl.load(x_ptr + offsets, mask=mask)
    output = tl.load(y_ptr + x.to(tl.int32), mask=mask)
```

full code and result: https://colab.research.google.com/drive/1fJjz7nMamxc66cwqF4nCuVpXDtM8mFUb?usp=drive_link

Actually, I also used indirect memory access and set `os.environ["TRITON_INTERPRET"] = "1"` in the same time in my previous code.

**Comment by wa008** [comment]
I print log and found `[tl.store](https://github.com/linkedin/Liger-Kernel/blob/bf48d8da1739bc01a2043e55e68697be1250e34c/src/liger_kernel/ops/cross_entropy.py#L255)` in cross_entropy doesn't work well

1) add `tl.device_print("device_print loss: ", loss)` at [L254](https://github.com/linkedin/Liger-Kernel/blob/bf48d8da1739bc01a2043e55e68697be1250e34c/src/liger_kernel/ops/cross_entropy.py#L254), output: 
```
(0, 0, 0)  device_print loss:  [0.42029184]
(1, 0, 0)  device_print loss:  [1.6193843]
(2, 0, 0)  device_print loss:  [0.8301048]
(3, 0, 0)  device_print loss:  [1.3625644]
```
Loss is correct.

2) add `print ("loss_1d after kernel: ", loss_1d)` at [L349](https://github.com/linkedin/Liger-Kernel/blob/bf48d8da1739bc01a2043e55e68697be1250e34c/src/liger_kernel/ops/cross_entropy.py#L349), output:
```
loss_1d after kernel:  tensor([0., 0., 0., 0.], device='cuda:0')
```

didn't know why
